falcon/api.py

```diff
@@ -244,12 +247,16 @@ def __call__(self, env, start_response):  # noqa: C901
                     # route, a default responder is returned and the
                     # resource is None. In that case, we skip the
                     # resource middleware methods.
-                    if resource is not None:
+                    if resource is not None and not resp.complete:
```

Comment:
Same here, I could remove `and not resp.complete` and still get everything to pass.

Reply:
In this case, when `resp.complete == True`, `resource` will always be `None`, which is why the check is superfluous. I went ahead and removed it and updated the comment.